core/src/main/scala/org/apache/spark/ui/jobs/StagePage.scala

```diff
@@ -786,9 +786,13 @@ private[spark] object ApiHelper {
     stageData.accumulatorUpdates.exists { acc => acc.name != null && acc.value != null }
   }
 
-  def hasInput(stageData: StageData): Boolean = stageData.inputBytes > 0
+  def hasInput(stageData: StageData): Boolean = {
+    stageData.inputBytes > 0 || stageData.inputRecords > 0
```

Comment:
Does it mean we didn't count the correct metrics somewhere?

Reply:
Yes, I think so. Anyway, it's better to determine based on either conditions as some input source might only have one metrics. I'll try to figure out why kafka source has such metrics in separate PR.

Reply:
If we're sure that it's because some resources do not provide certain metrics rather than they counted incorrectly, then, I think the current fix should be fine. Otherwise, I think we should fix the counting of metrics instead(as it's a bug).
Good to see to figure out the root cause first.

Reply:
I think root cause is as below, https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DataSourceRDD.scala#L96:
```
private class MetricsHandler extends Logging with Serializable {
  private val inputMetrics = TaskContext.get().taskMetrics().inputMetrics
  private val startingBytesRead = inputMetrics.bytesRead
  private val getBytesRead = SparkHadoopUtil.get.getFSBytesReadOnThreadCallback()

  def updateMetrics(numRows: Int, force: Boolean = false): Unit = {
    inputMetrics.incRecordsRead(numRows)
    val shouldUpdateBytesRead =
      inputMetrics.recordsRead % SparkHadoopUtil.UPDATE_INPUT_METRICS_INTERVAL_RECORDS == 0
    if (shouldUpdateBytesRead || force) {
      inputMetrics.setBytesRead(startingBytesRead + getBytesRead())
    }
  }
}
```
`DataSourceRDD` only supports updating records from hadoop read

Reply:
I see. Is it possible to add a UT for the change? If not, could you paste the screenshot for the UI change?

Reply:
Added screenshot